management/branding.go

```diff
@@ -16,10 +20,106 @@ type Branding struct {
 type BrandingColors struct {
 	// Accent color
 	Primary *string `json:"primary,omitempty"`
+
 	// Page background color
+	// Only one of PageBackground and PageBackgroundGradient should be set.
+	// If both fields are set, PageBackground takes priority.
+	PageBackground *string `json:"-"`
+
+	// Page background gradient
+	// Only one of PageBackground and PageBackgroundGradient should be set.
+	// If both fields are set, PageBackground takes priority.
+	PageBackgroundGradient *BrandingPageBackgroundGradient `json:"-"`
+}
+
+type BrandingPageBackgroundGradient struct {
+	Type        *string `json:"type,omitempty"`
+	Start       *string `json:"start,omitempty"`
+	End         *string `json:"end,omitempty"`
+	AngleDegree *int    `json:"angle_deg,omitempty"`
+}
+
+type jsonPageBackgroundSolid struct {
 	PageBackground *string `json:"page_background,omitempty"`
 }
 
+type jsonPageBackgroundGradient struct {
+	PageBackgroundGradient *BrandingPageBackgroundGradient `json:"page_background,omitempty"`
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+//
+// It is required to handle the json field page_background, which can either
+// be a hex color string, or an object describing a gradient.
+func (bc *BrandingColors) MarshalJSON() ([]byte, error) {
+	var data interface{} = struct {
+		Primary *string `json:"primary,omitempty"`
+	}{
+		Primary: bc.Primary,
+	}
+
+	if bc.PageBackground != nil {
+		data = struct {
+			Primary *string `json:"primary,omitempty"`
+			*jsonPageBackgroundSolid
+		}{
+			Primary: bc.Primary,
+			jsonPageBackgroundSolid: &jsonPageBackgroundSolid{
+				PageBackground: bc.PageBackground,
+			},
+		}
+	} else if bc.PageBackgroundGradient != nil {
+		data = struct {
+			Primary *string `json:"primary,omitempty"`
+			*jsonPageBackgroundGradient
+		}{
+			Primary: bc.Primary,
+			jsonPageBackgroundGradient: &jsonPageBackgroundGradient{
+				PageBackgroundGradient: bc.PageBackgroundGradient,
+			},
+		}
+	}
+
+	return json.Marshal(data)
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+//
+// It is required to handle the json field page_background, which can either
+// be a hex color string, or an object describing a gradient.
+func (bc *BrandingColors) UnmarshalJSON(data []byte) error {
+	// Unmarshal common fields
+	type JSONBrandingColors BrandingColors
+
+	var commonBrandingColors JSONBrandingColors
+
+	err := json.Unmarshal(data, &commonBrandingColors)
+	if err != nil {
+		return err
+	}
+
+	bc.Primary = commonBrandingColors.Primary
+
+	// Unmarshal page_background
+	// Can be either a string with solid color
+	var solidBrandingColor jsonPageBackgroundSolid
+	err = json.Unmarshal(data, &solidBrandingColor)
+	if err == nil {
+		bc.PageBackground = solidBrandingColor.PageBackground
+		bc.PageBackgroundGradient = nil
+	}
```

Comment:
Instead of trying out which Unmarshal doesn't return errors, we should unmarshal the field in question into an intermediate `interface{}` and from its type we can do what we need to do.

I'd call this field `RawPageBackground` of type `interface{}`, local only to the `MarshalJSON` or `UnmarshalJSON` methods. This would have the appropriate json tag (`json:"page_background,omitempty"`).

With this field populated, an assertion would be sufficient to tell us which kind it is. [This example](https://github.com/go-auth0/auth0/blob/master/management/user.go#L118-L129) does exactly that.

Reply:
The example looks really good.

After some issues with the type switch and `map[string]interface{}` (I forgot to add the `{}`...), I got this working. To convert from the `map[string]interface{}` to `BrandingPageBackgroundGradient` I first converted back to JSON and then into the correct struct.